src/broadcast.jl

```diff
@@ -102,15 +102,17 @@ end
         end
     end
 
-    eltype_exprs = [t <: AbstractArray ? :($(eltype(t))) : :($t) for t ∈ a]
-    newtype_expr = :(Core.Inference.return_type(f, Tuple{$(eltype_exprs...)}))
-
     return quote
         @_inline_meta
-        @inbounds return similar_type($first_staticarray, $newtype_expr, Size($newsize))(tuple($(exprs...)))
+        elements = tuple($(exprs...))
+        @inbounds return similar_type($first_staticarray, eltype(elements), Size($newsize))(elements)
```

Comment:
`eltype(::Tuple)` is (or at least, was) a bit pesimised. In `util.jl` there is a function called `promote_tuple_eltype` that I originally wrote for this kind of stuff. I'm not sure what the current status on which one is better (and it's possible that reducing via `typejoin` is a better choice here than `promote_type` anyway??)

Reply:
Interesting.  I didn't give type reduction a lot of thought TBH, because I'm expecting all the elements to have the same inferred type (How could they not? They're all copies of the same expression unrolled many times.)

In the known bad cases which required the broadcast workaround, the issue seems to be in inferring the type of the element tuple, rather than in the `typejoin` step.